[Tiang Hui Zheng] iP

heyzec opened the pull request:
## DukePro

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~FAST~ _SUPER_ FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2223s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :wink:

And it is **FREE**!

Features:
- [x] Managing tasks
- [x] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

___
If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
feat: add Gradle support

heyzec committed:
feat: customise chatbot from Duke to Hazell

heyzec committed:
feat: add skeletal code to echo and exit

heyzec committed:
feat: add reply function for formatting of replies

heyzec committed:
feat: add storing of text

heyzec committed:
refactor: create Task class

heyzec committed:
feat: add ability to mark tasks as done/undone

heyzec committed:
refactor: renamed methods and fields of Task class

heyzec committed:
refactor: make Task abstract and create ToDo class

heyzec committed:
feat: create Event, Deadline functionality

heyzec committed:
refactor: create Command class

- makes working with userinput easier by providing helper methods

heyzec committed:
refactor: create TaskStore class to store Tasks

heyzec committed:
fix: crash and no response

- crash: when marking task out of bounds
- no response: when command invalid
- empty response: when listing empty store

heyzec committed:
fix: adding tasks now shows detailed info of task

heyzec committed:
fix: change .gitignore, remove EXPECTED-UNIX.TXT

- it does not make sense that EXPECTED.TXT is not ignored but
  EXPECTED-UNIX.TXT is
- corrected spelling of ACTUAL.txt to ACTUAL.TXT

heyzec committed:
test: add semi-automated testing using I/O redir

heyzec committed:
feat: add TaskDescriptionEmpty, UnknownCommand exc

heyzec committed:
feat: KwargNotFoundEx is now a HazellException

heyzec committed:
refactor: new NoSuchTask handle Task out of bounds

heyzec committed:
feat: add deletion of tasks functionality

heyzec committed:
feat: friendlier response when listing tasks

heyzec committed:
refactor: use static factory over constructors

- in preparation for Level 7

heyzec committed:
refactor: make constructors not throw exceptions

heyzec committed:
refactor: make actions all done via TaskStore

- also in preparation for Level 7

heyzec committed:
feat: add saving and restoring tasks to file

heyzec committed:
fix: only recognise kwargs when space precedes "/"

heyzec committed:
feat: add ability to parse dates for deadlines

heyzec committed:
Merge branch 'branch-Level-7'

heyzec committed:
Merge branch 'branch-Level-8'

heyzec committed:
fix: use .equals string comparison & code cleanup

heyzec committed:
refactor: rename TaskStore with TaskList

heyzec committed:
fix: update testing to run twice

- to test if data is persisted on storage

heyzec committed:
refactor: create Ui class

heyzec committed:
refactor: create Storage class

heyzec committed:
refactor: mv files to packages, remove Maven conv

heyzec committed:
refactor: add folders to match Gradle convention

heyzec committed:
fix: remove .gradle files

heyzec committed:
test: added two basic unit tests

heyzec committed:
fix: update build.gradle file

heyzec committed:
docs: add more javadocs to classes and methods

heyzec committed:
style: comply with coding standards

heyzec committed:
feat: add search feature by keyword

heyzec committed:
Merge branch 'branch-A-JavaDoc'

heyzec committed:
Merge branch 'branch-A-CodingStandard'

heyzec committed:
Merge branch 'branch-Level-9'

midnightfeverrr commented on src/main/java/hazell/Command.java:
```diff
@@ -0,0 +1,76 @@
+package hazell;
+
+import hazell.exceptions.KwargNotFound;
+import java.util.*;
```
Avoid using asterisk to import (better to import explicitly)

midnightfeverrr commented on src/main/java/hazell/Command.java:
```diff
@@ -0,0 +1,76 @@
+        for (int i = 0; i < splitBySlash.length - 1; i++) {
+            String[] kwargWords = splitBySlash[i + 1].split(" ", 2);
+            kwargs.put(kwargWords[0], kwargWords[1]);
+        }
```
Very good implementations !

midnightfeverrr commented on src/main/java/hazell/Command.java:
```diff
@@ -0,0 +1,76 @@
+
+    /**
+     * Checks whether the first arg matches `word`.
+     * @param word The first word in the command.
```
Comment should have empty space between description of the method and the parameters.

midnightfeverrr commented on src/main/java/hazell/Deadline.java:
```diff
@@ -0,0 +1,53 @@
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+
+public class Deadline extends Task {
```
Missing javaDocs.

midnightfeverrr commented on src/main/java/hazell/Task.java:
```diff
@@ -0,0 +1,76 @@
+/**
+ * A superclass to be inherited by other Tasks.
+ */
+public abstract class Task {
```
Great OOP Concept of abstract class!

midnightfeverrr commented on src/main/java/hazell/TaskList.java:
```diff
@@ -0,0 +1,187 @@
+ * Class that stores Tasks.
+ */
+public class TaskList {
+    private final List<Task> store;
```
constants should be name with all capital letters. (i.e. STORE)

jialatteo commented on src/main/java/hazell/Event.java:
```diff
@@ -0,0 +1,39 @@
+package hazell;
+
+import hazell.exceptions.TaskDescriptionEmpty;
+
+public class Event extends Task {
```
Add JavaDocs to this class here?

jialatteo commented on src/main/java/hazell/Hazell.java:
```diff
@@ -0,0 +1,95 @@
+                } else if (command.startsWith("list")) {
+                    ui.reply(String.format("Here are the tasks in your list:\n%s", taskList));
+                } else if (command.startsWith("mark")) {
+                    int index = Integer.parseInt(command.getTrailingArgs().get(0)) - 1;
+                    String response = taskList.markTaskAsDone(index);
+                    ui.reply(response);
+                } else if (command.startsWith("unmark")) {
+                    int index = Integer.parseInt(command.getTrailingArgs().get(0)) - 1;
+                    String response = taskList.markTaskAsUndone(index);
+                    ui.reply(response);
+                } else if (command.startsWith("todo")) {
+                    String description = String.join(" ", command.getTrailingArgs());
+                    String response = taskList.addTask(ToDo.create(description));
+                    ui.reply(response);
+                } else if (command.startsWith("deadline")) {
+                    String description = String.join(" ", command.getTrailingArgs());
+                    String time = command.getKwarg("by");
+                    String response = taskList.addTask(Deadline.create(description, time));
+                    ui.reply(response);
+                } else if (command.startsWith("event")) {
+                    String description = String.join(" ", command.getTrailingArgs());
+                    String time = command.getKwarg("at");
+                    String response = taskList.addTask(Event.create(description, time));
+                    ui.reply(response);
+                } else if (command.startsWith("delete")) {
+                    int index = Integer.parseInt(command.getTrailingArgs().get(0)) - 1;
+                    String response = taskList.deleteTask(index);
+                    ui.reply(response);
+                } else if (command.startsWith("find")) {
+                    String keyword = command.getTrailingArgs().get(0);
+                    String response = taskList.findMatchingTasks(keyword);
+                    ui.reply(response);
+                } else {
+                    throw new UnknownCommand();
+                }
+            } catch (HazellException ex) {
+                ui.reply(ex.toString());
+            }
+        }
+    }
```
Might be better to parse commands in a Parser class rather than in Main class?

jialatteo commented on src/main/java/hazell/Storage.java:
```diff
@@ -0,0 +1,60 @@
+    private Path path;
+
+    /**
+     * Returns a Storage object pointing to the file used to store Hazell data locally.
+     *
+     * Before doing so, it ensures that the folder exists.
+     * @throws IOException
+     */
+    public Storage(String filePath) throws IOException {
```
Perhaps can add the missing param documentation here?

jialatteo commented on src/main/java/hazell/Ui.java:
```diff
@@ -0,0 +1,46 @@
+        this.scanner = new Scanner(System.in);
+    }
+
+    /**
+     * Check if user has entered a new command.
+     * @return Whether user has entered a new command
+     */
+    public boolean hasNextCommand() {
+        return this.scanner.hasNextLine();
+    }
```
Maybe can consider rephrasing comments to match java coding standard (i.e. CHECKS if user has entered a new commmand.)

jialatteo reviewed (commented):
Overall nice work with the chatbot so far! 😁

heyzec committed:
feat: add GUI template code from tutorial

heyzec committed:
style: fix style issues found by midnightfeverrr

heyzec committed:
style: fix style issues found by jialatteo

heyzec committed:
refactor: Dispatcher class suggested by jialatteo

heyzec committed:
refactor: use FXML

heyzec committed:
feat: enable GUI to give proper responses

heyzec committed:
refactor: move some classes into two new packages

heyzec committed:
style: fix issues with docs, remove dead code

heyzec committed:
Merge branch 'branch-Level-10'

heyzec committed:
feat: add assertions to code

- this documents important assumptions that should hold at various points in the code
- ensures user reply is not blank
- ensures timestring formatting did not go wrong

heyzec committed:
refactor: move GUI classes into own package

- Classes in the source root was slightly messy
- Also in preparation for further refactoring to abstract Ui related
  logic out of Hazell main class

heyzec committed:
refactor: improve code quality

- Added more documentation to classes and methods
- Made some fields final
- Change a large if-else to switch-case stmt

heyzec committed:
Merge pull request #2 from heyzec/branch-A-Assertions

feat: add assertions to code

heyzec committed:
Merge branch 'master' into branch-A-CodeQuality

heyzec committed:
Merge pull request #3 from heyzec/branch-A-CodeQuality

refactor: improve code quality

heyzec committed:
refactor: create UiInterface between Hazell and Ui

heyzec committed:
refactor: create classes to hide CLI, GUI diffs

heyzec committed:
feat: enable postponing of Event, Deadline tasks

heyzec committed:
docs: add basic userguide

heyzec committed:
style: fix import orders

heyzec committed:
style: added files required for stylecheck

heyzec committed:
fix: remove data/ folder from git index

heyzec committed:
fix: tests

heyzec committed:
style: fix minor issues - newline, space, layout

heyzec committed:
docs: add and improve docs for public methods

heyzec committed:
build: fix build.gradle so that JAR file uses GUI

heyzec committed:
fix: stop app from crashing if empty after switch

heyzec committed:
fix: change error smiley face to more visible one

heyzec committed:
feat: fix Level-8 Dates and Times

heyzec committed:
Quickfix for some bugs:

- App crashes when inputs are invalid
- User guide not aligned with actual output of app
